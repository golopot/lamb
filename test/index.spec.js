const {parser, reduce} = require('../index')


const testcases = [
  '(λx{x})(y)', // basic application
  '(λx{λy{x}})(t)', // closure
  '(λx{λy{λz{x}}})(t)(s)', // double-closure
  '(λx{λy{x}})(t)(s)', // non-local variable
  '(λx{λx{x}})(t)(s)', // parameter name collision
  '(λx{λt{x}})(t)(s)', // capture avoid
  '(λx{x})((λx{x})(t))', // reducible argument
  '(λx{λy{x(y)}})(t)(s)', // call in body
  '(λx{λy{x(y)}})(λx{x})(s)', // higher-order function
]

const negativeCases = [
  '(λx{x})y'
]

describe('parser', () => {
  for (let testcase of testcases){
    test(testcase, () => {
      const tree = parser.parse(testcase)
      expect(tree).toMatchSnapshot()
    })
  }
})

describe('reduce', () => {
  for (let testcase of testcases){
    test(testcase, () => {
      const tree = reduce(parser.parse(testcase))
      expect(tree).toMatchSnapshot()
    })
  }
})

describe('parser negative', () => {
  for (let negCase of negativeCases){
    test(negCase, () => {
      expect(() => {
        parser.parse(negCase)
      }).toThrow()
    })
  }
})
