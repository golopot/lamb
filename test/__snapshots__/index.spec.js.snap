// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser (λx{x})((λx{x})(t)) 1`] = `
Object {
  "argument": Object {
    "argument": "t",
    "func": Object {
      "type": "paren",
      "x": Object {
        "body": "x",
        "parameter": "x",
        "type": "λ",
      },
    },
    "type": "call",
  },
  "func": Object {
    "type": "paren",
    "x": Object {
      "body": "x",
      "parameter": "x",
      "type": "λ",
    },
  },
  "type": "call",
}
`;

exports[`parser (λx{x})(y) 1`] = `
Object {
  "argument": "y",
  "func": Object {
    "type": "paren",
    "x": Object {
      "body": "x",
      "parameter": "x",
      "type": "λ",
    },
  },
  "type": "call",
}
`;

exports[`parser (λx{λt{x}})(t)(s) 1`] = `
Object {
  "argument": "s",
  "func": Object {
    "argument": "t",
    "func": Object {
      "type": "paren",
      "x": Object {
        "body": Object {
          "body": "x",
          "parameter": "t",
          "type": "λ",
        },
        "parameter": "x",
        "type": "λ",
      },
    },
    "type": "call",
  },
  "type": "call",
}
`;

exports[`parser (λx{λx{x}})(t)(s) 1`] = `
Object {
  "argument": "s",
  "func": Object {
    "argument": "t",
    "func": Object {
      "type": "paren",
      "x": Object {
        "body": Object {
          "body": "x",
          "parameter": "x",
          "type": "λ",
        },
        "parameter": "x",
        "type": "λ",
      },
    },
    "type": "call",
  },
  "type": "call",
}
`;

exports[`parser (λx{λy{x(y)}})(t)(s) 1`] = `
Object {
  "argument": "s",
  "func": Object {
    "argument": "t",
    "func": Object {
      "type": "paren",
      "x": Object {
        "body": Object {
          "body": Object {
            "argument": "y",
            "func": "x",
            "type": "call",
          },
          "parameter": "y",
          "type": "λ",
        },
        "parameter": "x",
        "type": "λ",
      },
    },
    "type": "call",
  },
  "type": "call",
}
`;

exports[`parser (λx{λy{x}})(t) 1`] = `
Object {
  "argument": "t",
  "func": Object {
    "type": "paren",
    "x": Object {
      "body": Object {
        "body": "x",
        "parameter": "y",
        "type": "λ",
      },
      "parameter": "x",
      "type": "λ",
    },
  },
  "type": "call",
}
`;

exports[`parser (λx{λy{x}})(t)(s) 1`] = `
Object {
  "argument": "s",
  "func": Object {
    "argument": "t",
    "func": Object {
      "type": "paren",
      "x": Object {
        "body": Object {
          "body": "x",
          "parameter": "y",
          "type": "λ",
        },
        "parameter": "x",
        "type": "λ",
      },
    },
    "type": "call",
  },
  "type": "call",
}
`;

exports[`reduce (λx{x})((λx{x})(t)) 1`] = `
Object {
  "argument": "t",
  "func": Object {
    "type": "paren",
    "x": Object {
      "body": "x",
      "parameter": "x",
      "type": "λ",
    },
  },
  "type": "call",
}
`;

exports[`reduce (λx{x})(y) 1`] = `"y"`;

exports[`reduce (λx{λt{x}})(t)(s) 1`] = `"t"`;

exports[`reduce (λx{λx{x}})(t)(s) 1`] = `"s"`;

exports[`reduce (λx{λy{x(y)}})(t)(s) 1`] = `
Object {
  "argument": "y",
  "func": "x",
  "type": "call",
}
`;

exports[`reduce (λx{λy{x}})(t) 1`] = `
Object {
  "body": "x",
  "env": Object {
    "x": "t",
  },
  "parameter": "y",
  "type": "λ",
}
`;

exports[`reduce (λx{λy{x}})(t)(s) 1`] = `"t"`;
